### PR TITLE
use TypeArithmetic trait in cumsum! implementation

### PR DESCRIPTION
**Pull request by jw3126**
very minor refactoring.

### CODE REVIEW

File: base/multidimensional.jl [hunk 1]
```diff
@@ -574,12 +574,13 @@ function accumulate_pairwise(op, v::AbstractVector{T}) where T
 end
 
 function cumsum!(out, v::AbstractVector, axis::Integer=1)
-    # for types prone to numerical stability issues, we want
```
Comment (Sacha0) [inline base/multidimensional.jl]: This comment seems worth preserving in some form?

File: base/multidimensional.jl [hunk 2]
```diff
@@ -574,12 +574,13 @@ function accumulate_pairwise(op, v::AbstractVector{T}) where T
...
-    # accumulate_pairwise.
-    axis == 1 ? accumulate_pairwise!(+, out, v) : copy!(out,v)
+    _cumsum!(out, v, axis, TypeArithmetic(eltype(out)))
 end
 
-function cumsum!(out, v::AbstractVector{<:Integer}, axis::Integer=1)
+function _cumsum!(out, v, axis, ::ArithmeticRounds)
+    axis == 1 ? accumulate_pairwise!(+, out, v) : copy!(out,v)
```
Comment (Sacha0) [inline base/multidimensional.jl]: Missing space between arguments to `copy!`, and likewise below?

**Comment by andreasnoack** [comment]
What is the motivation for this?

**Comment by jw3126** [comment]
It was a hack. I wanted to dispatch on possibility of rounding errors, when I did #18931. But that trait did not yet exist, so I dispatched on `AbstractVector{<:Integer}` vs `AbstractVector`. Yesterday I discovered that we have such a trait and happily used it.

File: base/multidimensional.jl [hunk 3]
```diff
@@ -574,13 +574,18 @@ function accumulate_pairwise(op, v::AbstractVector{T}) where T
...
+    # we dispatch on the possibility of numerical stability issues
+    _cumsum!(out, v, axis, TypeArithmetic(eltype(out)))
 end
 
-function cumsum!(out, v::AbstractVector{<:Integer}, axis::Integer=1)
-    axis == 1 ? accumulate!(+, out, v) : copy!(out,v)
+function _cumsum!(out, v, axis, ::ArithmeticRounds)
+    axis == 1 ? accumulate_pairwise!(+, out, v) : copy!(out, v)
+end
+function _cumsum(out, v, axis, ::ArithmeticUnknown)
```
Comment (Sacha0) [inline base/multidimensional.jl]: Missing `!`? If so, perhaps add a test covering this code path?

Comment (jw3126) [inline base/multidimensional.jl]: Right, fixed!

**Review by Sacha0 — APPROVED** [review]
generally lgtm, but someone more familiar with this code should have a look! :)

**Comment by tkelman** [comment]
should be squashed on merge, if not before

**Comment by Sacha0** [comment]
@andreasnoack, lgty? :)

**Comment by Sacha0** [comment]
A second approving review of this pull request would be great. But absent that, and absent objections, requests for time, or someone beating me to it, I plan to merge this tomorrow morning. Best!

**Review by timholy — APPROVED** [review]

**Comment by Sacha0** [comment]
Thanks @timholy! :)

**Pull request merged**